### ISSUE TITLE
[Extension]: Controlnet - Can't upload independent image on 'Image' tab

### ISSUE THREAD

**Issue by rcwatkins** [vladmandic/automatic#2760]
### Issue Description

"Upload Independent Control Image" checkbox does not show ui element to upload image. 

Clean install tested, up to date extension. The only reference to this issue I can find in the controlnet extension repo is [here](https://github.com/Mikubill/sd-webui-controlnet/issues/2520#issuecomment-1902468258). Verified that 1.7.0 A1111 does work.

![image](https://github.com/vladmandic/automatic/assets/115605038/e79bf416-b770-4e0d-946c-26af716691e3)

### Version Platform Description

Starting SD.Next
Logger: file="C:\Stable Diffusion\Vlad\automatic\sdnext.log" level=INFO size=65 mode=append
Python 3.10.9 on Windows
Version: app=sd.next updated=2024-01-13 hash=5fb290f4 url=https://github.com/vladmandic/automatic/tree/master
Platform: arch=AMD64 cpu=Intel64 Family 6 Model 60 Stepping 3, GenuineIntel system=Windows release=Windows-10-10.0.19045-SP0 python=3.10.9
nVidia CUDA toolkit detected: nvidia-smi present

Commit: 5fb290f443d5f38a5c9f6e6095aabeab8e3a991d
OS: Windows 10
Browsers: Chrome, Firefox, Edge
GPU: GTX 1080

### URL link of the extension

https://github.com/Mikubill/sd-webui-controlnet

### URL link of the issue reported in the extension repository

https://github.com/Mikubill/sd-webui-controlnet/issues/2520#issuecomment-1902466367

### Acknowledgements

- [X] I have read the above and searched for existing issues

**Comment by dan4ik94** [comment]
I got the same issue.

![chrome_GiDwAxIvd1](https://github.com/vladmandic/automatic/assets/84623377/f46422d5-22de-4a0d-bacb-4ebb763e5f8e)

**Comment by nthmetaljustice** [comment]
Also seems to affect "Allow Preview" setting in Text2Img tab, where the preview pane doesn't show up anymore next to the controlnet image.

**Comment by TimothyKrell** [comment]
I also noticed when selecting a preprocessor, no controls are show (like when selecting tile, there is no slider for blurring).

**Comment by vladmandic** [comment]
issue was introduced in controlnet via [commit](https://github.com/Mikubill/sd-webui-controlnet/commit/b146963c1b0067dcea478a0ff436431f3328c1db) which has hard-coded variables for a1111 v1.7.0 only

see original [issue](https://github.com/Mikubill/sd-webui-controlnet/issues/2516)

subsequent [hotfix](https://github.com/Mikubill/sd-webui-controlnet/pull/2518) makes it compatible with a1111 v1.6.0 as well, but nothing else

until this is resolved upstream, i'm locking controlnet version to commit `ecd33eb`

if you run `webui --upgrade`, it will automatically adjust to last compatible version

cc: @huchenlei

**Comment by huchenlei** [comment]
Hmmm. Sorry about inconvenience. The intention for this PR is to initialize ControlNet after every components expected by ControlNet in the main UI has appeared.
Do you have any idea on a component that is rendered last on the UI that we can anchor to? or do you have better solution on this problem?

**Comment by vladmandic** [comment]
there is a bool flag for that, just check `ui.loadsave.finalized_ui`

**Comment by huchenlei** [comment]
I don't think that will work, as you have no way to access that boolean flag when the UI is constructing. `UiLoadsave` object is bind to `shared.demo.loadsave`  after all gradio elements are initialized.

The core problem is how can we find a way to register a callback when UI is fully ready. `modules/script_callbacks` simply does not provide that hook.

**Comment by huchenlei** [comment]
I will change the extension to disable callback on that checkpoint settings for now so that SDNext users can be unblocked.

**Comment by vladmandic** [comment]
a) i don't mind providing a callback, but you'd have to add same to a1111 unless you want to have different implementation
b) wrapping access to `modules.ui.loadsave.finalized_ui` in a try/except getter is easy

**Comment by huchenlei** [comment]
```python
            loadsave.setup_ui() <-- finalized_ui flag is set to True here

        if os.path.exists(os.path.join(script_path, "notification.mp3")) and shared.opts.notification_audio:
            gr.Audio(interactive=False, value=os.path.join(script_path, "notification.mp3"), elem_id="audio_notification", visible=False)

        footer = shared.html("footer.html")
        footer = footer.format(versions=versions_html(), api_docs="/docs" if shared.cmd_opts.api else "https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/API")
        gr.HTML(footer, elem_id="footer")

        settings.add_functionality(demo)

        update_image_cfg_scale_visibility = lambda: gr.update(visible=shared.sd_model and shared.sd_model.cond_stage_key == "edit")
        settings.text_settings.change(fn=update_image_cfg_scale_visibility, inputs=[], outputs=[image_cfg_scale])
        demo.load(fn=update_image_cfg_scale_visibility, inputs=[], outputs=[image_cfg_scale])

        modelmerger_ui.setup_ui(dummy_component=dummy_component, sd_model_checkpoint_component=settings.component_dict['sd_model_checkpoint'])

    loadsave.dump_defaults()
    demo.ui_loadsave = loadsave <-- loadsave is bind to demo here
   
    return demo <-- no more gradio create component logic
```
So in the callback I currently use, i.e. `on_after_component` won't trigger after `modules.ui.loadsave.finalized_ui` becomes available.

**Comment by vladmandic** [comment]
i've added `on_after_ui` callback if you want to use it (but should check if it exists since a1111 does not have it nor do older versions of sdnext)
